clang/lib/CodeGen/CGOpenMPRuntime.cpp

```diff
@@ -7731,10 +7731,30 @@ class MappableExprsHandler {
               IsImplicit, Mapper, VarRef, ForDeviceAddr);
         };
 
+    // Sort all map clauses and make sure all the maps containing array
+    // sections are processed last.
+    llvm::SmallVector<const OMPMapClause *, 16> SortedMapClauses;
```

Comment:
What implies that this is sorted?

Reply:
I don't understand the question.

Reply:
Are you asking what is the sorting criteria?

Reply:
It's called "SortedMapClauses" but I don't see any sorting, we just push back into the vector.

Reply:
Ah yes, so I just moved all the maps containing sections at the end of the clause list. I want those maps to happen last after all the structs and other maps have happened.

Reply:
This sorting approach is not very good, it adds lots of the code which is not required. It was supposed that this function automatically adjusts the boundaries of the structs/classes, so sorting is not required.

Reply:
It's a form of sorting it's more like a split between all section-containing maps and the ones that don't.

Reply:
@alexey-bataev I agree it's not ideal. the problem is related to the order in which the clauses are processed. We cannot process the base struct after we have processed an array section inside the struct.

Reply:
Why, what's wrong with this?

Reply:
at runtime, if things happen in the wrong order, the processing of the base struct overwrites the pointer attachment for the array.

Reply:
But that what this code already does, it (implicitly) "reorders" the clauses by extending the mapped region automatically. Explicit reordering should not be required.

Reply:
Implicit ordering isn't working in the case in the example above, please see the code. The entries are in the wrong order in the runtime and the problem starts here.

Reply:
And this is what should be fixed, no need to add a hack, that "hides" the bug in the code, need to fix the bug.

Reply:
Well I don't see anything other that's wrong other than the order and the order comes from how the user wrote the code so I am not sure how else to fix it.

Reply:
I can't find any "bug" in the existing code. It works as intended. The problem is that it doesn't handle these types of situations and I don't see how else to fix an ordering problem other than by re-ordering. If you have a different solution in mind please let me know.

Reply:
@alexey-bataev I have looked at the code again and I really can't see another solution to this problem. If you have a different fix in mind please let me know.

Reply:
Will try to investigate it tomorrow, probably